docs/modules/ROOT/pages/guides/src5-migration.adoc

```diff
@@ -0,0 +1,97 @@
+SNIP-5 allows interface ID calculations to use Cairo types and the Starknet keccak (`sn_keccak`) function.
+For more information on the decision, see the {shamans-proposal} or the {dual-interface-discussion} discussion.
+
+== Migrating to SRC5
+
+:src5-component: xref:api/introspection#SRC5Component[SRC5Component]
+:upgradeable-component: xref:api/upgrades#UpgradeableComponent[UpgradeableComponent]
+:isrc6: xref:api/account.adoc#ISRC6[ISRC6]
+:src5-rs: https://github.com/ericnordelo/src5-rs[src5-rs]
+
+This guide is for already-deployed, upgradeable contracts that support ERC165.
+Migrating to SRC5 requires integrating {src5-component} and {upgradeable-component}.
+The following template assumes that the upgradeable mechanism uses the `upgrade` selector:
+
+[,javascript]
+----
+#[starknet::contract]
+mod MigratingContract {
+    use openzeppelin::introspection::src5::SRC5Component;
+    use openzeppelin::upgrades::UpgradeableComponent;
+    use starknet::ClassHash;
+    use starknet::ContractAddress;
+
+    component!(path: SRC5Component, storage: src5, event: SRC5Event);
+    component!(path: UpgradeableComponent, storage: upgradeable, event: UpgradeableEvent);
+
+    // SRC5
+    #[abi(embed_v0)]
+    impl SRC5Impl = SRC5Component::SRC5Impl<ContractState>;
+    #[abi(embed_v0)]
+    impl SRC5CamelOnlyImpl = SRC5Component::SRC5CamelImpl<ContractState>;
+    impl SRC5InternalImpl = SRC5Component::InternalImpl<ContractState>;
+
+    // Upgradeable
+    impl UpgradeableInternalImpl = UpgradeableComponent::InternalImpl<ContractState>;
+
+    #[storage]
+    struct Storage {
+        #[substorage(v0)]
+        src5: SRC5Component::Storage,
+        #[substorage(v0)]
+        upgradeable: UpgradeableComponent::Storage,
+    }
+
+    #[event]
+    #[derive(Drop, starknet::Event)]
+    enum Event {
+        #[flat]
+        SRC5Event: SRC5Component::Event,
+        #[flat]
+        UpgradeableEvent: UpgradeableComponent::Event
+    }
+
+    #[external(v0)]
+    fn upgrade(ref self: ContractState, new_class_hash: ClassHash) {
+        // Add permissions mechanism
+        self.upgradeable._upgrade(new_class_hash);
+    }
+
+    // The rest of the contract logic
+}
+----
+
+The contract will also need to register the new SRC5 interface IDs in order to continue declaring support for those interfaces.
+We can create a migration initializer to handle this which needs to be called after the contract upgrade.
+
+TIP: OpenZeppelin's Contracts for Cairo already provides the interface IDs of common interfaces in the API section (like {isrc6}).
+For interfaces not provided, tools such as {src5-rs} can help with interface ID calculation.
+
+[,javascript]
+----
+#[starknet::contract]
+mod MigratingContract {
+
+    (...)
+
+    #[external(v0)]
+    fn migration_initializer(ref self: ContractState, interface_ids: Span<felt252>) {
+        // Add permissions mechanism
+        self.src5.register_interfaces(interface_ids);
+    }
+}
```

Comment:
The focus of the guide should be this migration function. Some comments:

* we should add minimal access control, or maybe just comments to make it shorter
* we should also add some anti-reinitialization check
* it should also deregister preexisting ones
* I'm honestly not convinced about the new `register_interfaces`. somewhat interface-conservative + complexity minimization bias, and I'm unsure I see the value in it. In this example, I would hardcode the ids in the code instead of taking them as arguments. You should know the ids by the time you write the contract.

Reply:
> we should add minimal access control, or maybe just comments to make it shorter
we should also add some anti-reinitialization check
it should also deregister preexisting ones

Done!

> I'm honestly not convinced about the new register_interfaces. somewhat interface-conservative + complexity minimization bias, and I'm unsure I see the value in it.

That's fair. I was trying to make it as easy as possible to migrate, but I agree that it is unnecessary. I did, however, add `deregister_erc165_interface` to the `utils` module. IMO this is a necessary one to provide. Let me know what you think

Reply:
i know i said we should deregister but is it really worth it, if there's no `supportsInterface` one looking into `ERC165_supported_interfaces`? unless there's any noticeable data/gas savings, it's probably not needed